cpp/src/strings/find.cu

```diff
@@ -0,0 +1,333 @@
+    {
+        CUDF_EXPECTS( !target.empty(), "Parameter target must not be empty.");
+        CUDF_EXPECTS( start >= 0, "Parameter start must be positive integer or zero.");
+        if( (stop) > 0 && (start >stop) )
+            CUDF_FAIL( "Parameter start must be less than stop.");
+        //
+        auto target_ptr = cudf::strings::detail::string_from_host(target.c_str(), stream);
+        auto d_target = *target_ptr;
+        auto strings_column = cudf::column_device_view::create(strings.parent(),stream);
+        auto d_column = *strings_column;
+        // copy the null mask
+        auto strings_count = strings.size();
+        rmm::device_buffer null_mask;
+        cudf::size_type null_count = d_column.null_count();
+        if( d_column.nullable() )
+            null_mask = rmm::device_buffer( d_column.null_mask(),
+                                            cudf::bitmask_allocation_size_bytes(strings_count),
+                                            stream, mr);
+        // create output column
+        auto results = std::make_unique<cudf::column>( cudf::data_type{cudf::experimental::type_to_id<IntegerType>()},
+            strings_count, rmm::device_buffer(strings_count * sizeof(IntegerType), stream, mr),
+            null_mask, null_count);
+        auto results_view = results->mutable_view();
+        auto d_results = results_view.data<IntegerType>();
+        // set the position values by evaluating the passed function
+        thrust::transform( rmm::exec_policy(stream)->on(stream),
+            thrust::make_counting_iterator<cudf::size_type>(0),
+            thrust::make_counting_iterator<cudf::size_type>(strings_count),
+            d_results, find_fn<IntegerType,FindFunction>{d_column, pfn, d_target, start, stop});
+        //
+        results->set_null_count(null_count);
+        return results;
+    }
+
+    template <typename IntegerType, typename FindFunction, std::enable_if_t<not std::is_integral<IntegerType>::value>* = nullptr>
+    std::unique_ptr<cudf::column> operator()( cudf::strings_column_view const& strings,
+                                              std::string const& target,
+                                              cudf::size_type start, cudf::size_type stop,
+                                              FindFunction& pfn,
+                                              rmm::mr::device_memory_resource* mr,
+                                              cudaStream_t stream ) const
+    {
+        CUDF_FAIL("Output type must be integral type.");
+    }
+};
+
+} // namespace
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+std::unique_ptr<cudf::column> find( cudf::strings_column_view const& strings,
+                                    std::string const& target,
+                                    cudf::size_type start=0, cudf::size_type stop=-1,
+                                    cudf::data_type output_type = cudf::data_type{INT32},
```

Comment:
Is there a good reason to support output types other than INT32? Would simplify the code a bit to just always return INT32 (no type dispatch).

Reply:
I would be happy to support just one type if that is allowed.

Reply:
@kkraus14 For input on this.

Reply:
Also @jlowe @revans2 -- is there a need to support arbitrary integer output types for string find()/rfind() etc?

Reply:
I spoke to Keith the end of last week and he said we recommended just supporting INT32 for this.

Reply:
INT32 should be fine.  SQL uses 1-based offsets for strings, so Spark will have to adjust the result. The cast to the desired output type can occur during that +1 adjustment.